### PR TITLE
Closes #3217: Add anchor link plugin for CKEditor5.

### LINKED ISSUE #3217: Anchor links stripped by CKEditor5
## Problem/Motivation

Anchor links impossible to create within CK Editor 5

### Describe the bug

- Anchor link HTML <a name> and <a id> are being stripped out CK Editor 5.
- Using <a html> doesn't seem to work as an anchor link, though CKE5 does not strip out above code.
- Verified in Safari, FF and Chrome

### To Reproduce

[https://ge.arizona.edu/program-history#Test](https://ge.arizona.edu/program-history#Test) - uses HTML shown below:

Screenshot 2024-02-26 at 2 41 50 PM (1)

## Proposed resolution

Add [CKEditor Anchor Link](https://www.drupal.org/project/anchor_link) module. 

Joe says: Looks like the required library even exists on asset-packagist:
[https://asset-packagist.org/package/npm-asset/northernco--ckeditor5-anchor-drupal](https://asset-packagist.org/package/npm-asset/northernco--ckeditor5-anchor-drupal
)
### Expected behavior

New anchor link button (looks like a flag) allows user to add anchor link to a page. 

anchor-link-3-cmd-section-1--top--bottom

Anchor link functions as expected.

## Additional context

[https://www.drupal.org/project/anchor_link](
https://www.drupal.org/project/anchor_link)

[https://github.com/ckeditor/ckeditor5/issues/8733](https://github.com/ckeditor/ckeditor5/issues/8733)

[https://github.com/northernco/ckeditor5-anchor-drupal](https://github.com/northernco/ckeditor5-anchor-drupal)

### PR DESCRIPTION
**Pull request by joeparsons**
## Description
Adds [CKEditor5 anchor link module](https://www.drupal.org/project/anchor_link) (and [plugin](https://github.com/northernco/ckeditor5-anchor-drupal)).

## Related issues

Closes #3217

## How to test

- Add anchor links to a text area using the AZ Standard editor
- Save the content
- See if the editor stripped out the anchor links

### Possible release notes.
```
Users with the content editor role can now easily create anchor links in text areas using the Standard text formatter.
```

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [x] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 1]
```diff
@@ -83,3 +83,10 @@ function az_paragraphs_update_920601(&$sandbox) {
   $sandbox['#finished'] = ($sandbox['total'] === '0') ? 1 : ($sandbox['current'] / $sandbox['total']);
   return t('Paragraph behavior settings fixed on %count paragraphs.', ['%count' => $sandbox['updated_count']]);
 }
+
+/**
+ * Enable anchor_link module for CKEditor5 by default.
+ */
+function az_quickstart_update_1021002() {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:90]: Update number may need to change depending on whether we decide to include this in a patch release and/or backport.

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 2]
```diff
@@ -83,3 +83,10 @@ function az_paragraphs_update_920601(&$sandbox) {
   $sandbox['#finished'] = ($sandbox['total'] === '0') ? 1 : ($sandbox['current'] / $sandbox['total']);
   return t('Paragraph behavior settings fixed on %count paragraphs.', ['%count' => $sandbox['updated_count']]);
 }
+
+/**
+ * Enable anchor_link module for CKEditor5 by default.
+ */
+function az_paragraphs_update_1021002() {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:90]: This update function number might need to change depending on if we decide to include backport this change / include it in a patch release and depending on whether this PR is merged before or after #2482.

File: modules/custom/az_paragraphs/az_paragraphs.module [hunk 3]
```diff
@@ -53,3 +53,19 @@ function az_paragraphs_theme_registry_alter(&$theme_registry) {
     }
   }
 }
+
+/**
+ * Implements hook_library_info_alter().
+ *
+ * Alter path of CKE5 anchor link library to match the one installed by composer.
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.module:60]:
```suggestion
 * Alter path of CKE5 anchor link library to match the one installed by composer.
 *
 * @link https://www.drupal.org/project/anchor_link/issues/3406349#comment-15463373
```

**Review by natebyerley — APPROVED** [review]
I tested this on Page, News and Event.  Works. 

I created a Block with an anchor link in it, and then embedded block in an Event.  Anchor link worked as expected. 

Approved.

**Comment by danahertzberg** [comment]
Requested improvements
- Maybe an interim solution could be added to CKEditor?
- Look into changing icon for this in the CKEditor
- Look into changing styles for text with anchor links

**Comment by danahertzberg** [comment]
Looks like it adds the link styling even though the text is not clickable..

**Comment by joeparsons** [comment]
A beta version is now available.  We should update this PR to use that release.

File: composer.json [hunk 4]
```diff
@@ -38,6 +38,7 @@
         "composer/installers": "^2.0",
         "cweagans/composer-patches": "^1.7",
         "drupal/access_unpublished": "1.5.0",
+        "drupal/anchor_link": "3.0.0-alpha2",
```
Comment (joeparsons) [inline composer.json:41]:
```suggestion
        "drupal/anchor_link": "3.0.0-beta1",
```

**Comment by danahertzberg** [comment]
How can we add CSS to have style not match the link? Should we add some form of icon like [Bootstrap docs](https://digital.arizona.edu/arizona-bootstrap/docs/2.0/layout/grid/#responsive-classes) does?

**Comment by bberndt-uaz** [comment]
There is an existing issue for the module for changing the icon: [A flag icon is usually for reporting content, use a hash instead](https://www.drupal.org/project/anchor_link/issues/3445509).

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 5]
```diff
@@ -90,3 +90,10 @@ function az_paragraphs_update_920601(&$sandbox) {
 function az_paragraphs_update_1021001() {
   \Drupal::service('module_installer')->install(['ckeditor_indentblock']);
 }
+
+/**
+ * Enable anchor_link module for CKEditor5 by default.
+ */
+function az_paragraphs_update_1021002() {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:97]:
```suggestion
function az_paragraphs_update_1021201() {
```

File: modules/custom/az_paragraphs/css/az_paragraphs_ckeditor_anchor_icon.css [hunk 6]
```diff
@@ -0,0 +1,4 @@
+div[contenteditable=true] .ck-anchor,
+.ckeditor5-toolbar-button-anchor {
+  background: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='512' height='512' viewBox='120 -880 720 800' fill='#000000'><path d='M480-80q-61 0-125-22t-116-60q-52-38-85.5-89T120-360v-120l160 120-62 62q29 51 92 88t130 47v-357H320v-80h120v-47q-35-13-57.5-43.5T360-760q0-50 35-85t85-35q50 0 85 35t35 85q0 39-22.5 69.5T520-647v47h120v80H520v357q67-10 130-47t92-88l-62-62 160-120v120q0 58-33.5 109T721-162q-52 38-116 60T480-80Zm0-640q17 0 28.5-11.5T520-760q0-17-11.5-28.5T480-800q-17 0-28.5 11.5T440-760q0 17 11.5 28.5T480-720Z'/></svg>") no-repeat left center;
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/css/az_paragraphs_ckeditor_anchor_icon.css:3]:
```suggestion
  background: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="120 -880 720 800" fill="%23000000"><path d="M480-80q-61 0-125-22t-116-60q-52-38-85.5-89T120-360v-120l160 120-62 62q29 51 92 88t130 47v-357H320v-80h120v-47q-35-13-57.5-43.5T360-760q0-50 35-85t85-35q50 0 85 35t35 85q0 39-22.5 69.5T520-647v47h120v80H520v357q67-10 130-47t92-88l-62-62 160-120v120q0 58-33.5 109T721-162q-52 38-116 60T480-80Zm0-640q17 0 28.5-11.5T520-760q0-17-11.5-28.5T480-800q-17 0-28.5 11.5T440-760q0 17 11.5 28.5T480-720Z"/></svg>') no-repeat left center;
```

**Comment by joeparsons** [comment]
FWIW it looks like the plugin was _available_ in Quickstart 1 but not enabled by default:
Screenshot 2024-10-22 at 2 53 26 PM

Also worth nothing that the plugin that was available in Quickstart 1 used the flag icon.

File: modules/custom/az_paragraphs/az_paragraphs.module [hunk 7]
```diff
@@ -6,6 +6,7 @@
  */
 
 use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\ckeditor5\Plugin\CKEditor5PluginDefinition;
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.module:9]:
```suggestion
```

**Review by trackleft — APPROVED** [review]

**Review by ewlyman — APPROVED** [review]

**Review by danahertzberg — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**